### PR TITLE
Optimizations to solve() function and more performance tests.

### PR DESCRIPTION
**Pull request by sdquiring**
Runtime optimizations:

 - `solution->solve()` optimized by moving debug options to default off and reserving space to avoid reallocations.
 - PropertyMatching can now use a binary search and early loop exit based on the first dimension of the key.
 - PropertyMatching no longer uses the batch size as a dimension of its distance metric.
 - CachingLibrary can now keep track of statistics regarding hits & misses.
 - Raw performance of the PropertyMatching lookup has been increased by moving virtual calls to templated calls & reducing datatype conversions.

Developer focused optimizations:

 - RunGEMMKernelTest now caches CPU reference results which significantly speeds up running the HostLibraryTests.
 - Moved the ParallelMap Python function to a separate Parallel.py file.  It's still accessible via Common.ParallelMap.

### CODE REVIEW

File: HostLibraryTests/llvm/LibraryPerformance_test.cpp [hunk 1]
```diff
@@ -35,58 +34,156 @@
...
+            return iter->second;
+
+        return libraryCache[path] = LoadLibraryFile<ContractionProblem>(libraryPath());
+    }
 };
 
+std::map<std::string, std::shared_ptr<SolutionLibrary<ContractionProblem>>> LibraryPerformanceTest::libraryCache;
+
+TEST_P(LibraryPerformanceTest, PopulateCache)
+{
```
Comment (AlexBrownAMD) [inline HostLibraryTests/llvm/LibraryPerformance_test.cpp]: Is this test supposed to be empty?

File: Tensile/Source/lib/include/Tensile/PropertyMatching.hpp [hunk 2]
```diff
@@ -150,10 +140,207 @@ namespace Tensile
...
+            }
+
             ReturnValue findBestKeyMatch(Key const& key, Transform transform) const
             {
                 const bool debug = Debug::Instance().printPropertyEvaluation();
+                const bool naive = Debug::Instance().naivePropertySearch();
+
+                if(naive)
+                {
+                    if(debug)
```
Comment (AlexBrownAMD) [inline Tensile/Source/lib/include/Tensile/PropertyMatching.hpp]: Is debug set at runtime?

**Comment by wbgilmartin** [comment]
The rocblas problem that fails is:

- { rocblas_function: "rocblas_sgemm", transA: 'N', transB: 'N', M: 16328, N: 384, K: 384, lda: 16328, ldb: 125000, ldc: 125000 }

**Comment by sdquiring** [comment]
Do you mean an actual segfault or a GPU Memory access fault?

On Mar 4, 2020, at 12:15 PM, wbgilmartin <notifications@github.com<mailto:notifications@github.com>> wrote:

[CAUTION: External Email]

The rocblas problem that fails is:

  *   { rocblas_function: "rocblas_sgemm", transA: 'N', transB: 'N', M: 16328, N: 384, K: 384, lda: 16328, ldb: 125000, ldc: 125000 }

**Comment by wbgilmartin** [comment]
actually memory access fault.

Memory access fault by GPU node-4 (Agent handle: 0xe84100) on address 0x7f0ea6296000. Reason: Page not present or supervisor privilege.
Aborted (core dumped)

File: HostLibraryTests/llvm/LibraryPerformance_test.cpp (hunk 1 above)
Comment (sdquiring) [reply HostLibraryTests/llvm/LibraryPerformance_test.cpp]: Yes, see the comment above.  I'll add a comment in the body of the test to make it clearer.

File: Tensile/Source/lib/include/Tensile/PropertyMatching.hpp (hunk 2 above)
Comment (sdquiring) [reply Tensile/Source/lib/include/Tensile/PropertyMatching.hpp]: Yes, it's set by one of the TENSILE_DB bits.

**Comment by sdquiring** [comment]
I believe I've fixed the memory access fault.  The problem was with a discrepancy between the MagicDivAlg value between the assembly of the kernel and the value in the TensileLibrary.yaml file used by the host runtime.

**Comment by sdquiring** [comment]
The issues found in this branch are also present in develop.  Planning on merging this PR once latest pre-checkin CI passes.

**Pull request merged**